File: Sorting Algorithms/Gabri/Divide&Conquer/Merge sort/MergeSort.py
def merge(arrA, arrB, low, mid, high):
	if (high + 1 - low >= 0):
		for i in range(low, high+1):    
			arrB[i] = arrA[i]

	i = low
	j = mid + 1

	for k in range(low, high+1):
		if (i > mid):
			arrA[k] = arrB[j]
			j+=1
		elif (j > high):
			arrA[k] = arrB[i]
			i+=1
		elif (arrB[j] < arrB[i]):
			arrA[k] = arrB[j]
			j+=1
		else:
			arrA[k] = arrB[i]
			i+=1


def merge_sort_rec(arrA, arrB, low, high):
	if (high > low):
		mid = low + (high-low)//2
		
		merge_sort_rec(arrA, arrB, low, mid)
		merge_sort_rec(arrA, arrB, mid+1, high)
		
		merge(arrA, arrB, low, mid, high)


def merge_sort(arrA):
	arrBuffer = [0] * len(arrA)
	merge_sort_rec(arrA, arrBuffer, 0, len(arrA) - 1)


def main():
	numbers = [89, 45, 68, 90, 29, 34, 17]
	print(f'Unsorted: {numbers}')

	merge_sort(numbers)
	print(f'Sorted: {numbers}')

main()

File: Sorting Algorithms/Gabri/Divide&Conquer/Merge sort/MergeSort.kt
fun mergeSort(arrA: IntArray) {
    val arrBuffer = IntArray(arrA.size)
    mergeSortRecursive(arrA, arrBuffer, 0, arrA.size - 1)
}

fun mergeSortRecursive(arrA: IntArray, arrB: IntArray, low: Int, high: Int) {
    if (high > low) {
        val mid = low + (high - low) / 2

        mergeSortRecursive(arrA, arrB, low, mid)
        mergeSortRecursive(arrA, arrB, mid + 1, high)
        merge(arrA, arrB, low, mid, high)
    }
}

fun merge(arrA: IntArray, arrB: IntArray, low: Int, mid: Int, high: Int) {
    for (k in low..high) arrB[k] = arrA[k]

    var i = low
    var j = mid + 1

    for (k in low..high) {
        if (i > mid) arrA[k] = arrB[j++]
        else if (j > high) arrA[k] = arrB[i++]
        else if (arrB[j] < arrB[i]) arrA[k] = arrB[j++]
        else arrA[k] = arrB[i++]
    }
}

fun main() {
    val arrA = intArrayOf(89, 45, 68, 90, 29, 34, 17)
    println("Unordered: " + arrA.contentToString())

    mergeSort(arrA)
    println("Ordered: " + arrA.contentToString())
}